task-2/server.ts

```diff
@@ -0,0 +1,119 @@
+import express from 'express';
+import { v4 as uuidv4 } from 'uuid';
+import 'dotenv/config';
+import { User } from './interfaces/user';
+import { userCreateSchema, userUpdateSchema, passwordSchema } from './schemas/user.schema';
+import { errorResponse } from './helpers';
+import USERS from './data/users.json';
+
+const PORT = process.env.PORT;
+
+const app = express();
+const userRouter = express.Router();
+
+app.set('x-powered-by', false);
+app.use(express.json());
+app.use('/users', userRouter);
+
+
+userRouter.route('/')
+	.post((req, res) => {
+		const user: User = req.body;
+
+		const { error } = userCreateSchema.validate(user, {
+			abortEarly: false,
+			allowUnknown: false,
+		});
+		if (error?.isJoi) {
+			res.status(400).json(errorResponse(error.details));
+			return;
+		}
+
+		const passwordErrors = passwordSchema.validate(user.password, { list: true })
+		if (passwordErrors.length) {
+			res.status(400).json({ message: `password error: ${passwordErrors.join(', ')}` });
+			return;
+		}
+
+		const oldUser = USERS.find(existingUser => existingUser.login === user.login);
+		if (oldUser) {
+			res.status(404).json({ message: 'User with such login already exists' });
+			return;
+		}
+
+		const newUser = { ...user, id: uuidv4(), isDeleted: false };
+		USERS.push(newUser);
+		res.status(200).send(newUser);
+	})
+	.get((req, res) => {
+		const { limit, loginSubstring = '' } = req.query;
```

Comment:
`req.query`  has type `QueryString.ParsedQs`

should be `(req: ValidatedRequest<AutoSuggestUsersRequestSchema>, res: Response)` where `export interface AutoSuggestUsersRequestSchema extends ValidatedRequestSchema {
    [ContainerTypes.Query]: AutoSuggestUsersRequestQuery;
}` and `export interface AutoSuggestUsersRequestQuery {
    loginSubstring: string;
    limit: number;
}`

Reply:
where did you get `ValidatedRequest` and `ValidatedRequestSchema` from?

Reply:
this all from `express-joi-validation`;